src/pybind/mgr/dashboard/services/proto/CMakeLists.txt

```diff
@@ -0,0 +1,118 @@
+
+# This will generate the correct protoc versions based on the given OS system python3 grpc module.
+
+# Failing to find any module version it will just not generate anything but still continue the cmake build.
+
+# For any of the most recent gprc_tools it should just work provided there are no major differences in behaviour.
+
+# This implementation should also be able to cope with repo updates to the `src/nvmeof/gateway` module.
+
+
+
+#####################################
+# Notes about python package loading:
+#####################################
+
+# For`gateway_pb2_grpc.py` to be successfully imported by python
+# These files must be created from the dashboard parent directory
+# so that `gateway_pb2_grpc.py` is written with an import statement that matches the module location. e.g. import dashboard.services.proto.gateway_pb2
+
+# WORKING_DIRECTORY: /ceph/src/pybind/mgr/dashboard/services/proto/../../..
+# /usr/bin/python3 -m grpc_tools.protoc --proto_path=. --python_out=. --grpc_python_out=. dashboard/services/proto/gateway.proto
+
+# view from python:
+
+# <module 'dashboard.services.orchestrator' from '/ceph/src/pybind/mgr/dashboard/services/orchestrator.py'>,
+# <module 'dashboard.services.nvmeof_conf' from '/ceph/src/pybind/mgr/dashboard/services/nvmeof_conf.py'>,
+# <module 'dashboard.services.nvmeof_cli' from '/ceph/src/pybind/mgr/dashboard/services/nvmeof_cli.py'>,
+# <module 'dashboard.services.nvmeof_client' from '/ceph/src/pybind/mgr/dashboard/services/nvmeof_client.py'>,
+# <module 'dashboard.services.proto' (namespace) from ['/ceph/src/pybind/mgr/dashboard/services/proto']>,
+# <module 'dashboard.services.proto.gateway_pb2' from '/ceph/src/pybind/mgr/dashboard/services/proto/gateway_pb2.py'>,
+# <module 'dashboard.services.proto.gateway_pb2_grpc' from '/ceph/src/pybind/mgr/dashboard/services/proto/gateway_pb2_grpc.py'>]
+
+# The grpc_tool generates an import statement for the gateway_pb2_grpc.py to the gateway_pb2.py based on the location of the file.
+# This import statement is static and not relative and you have no control over it other than by modifying their path at generation time from the tool.
+
+# By explicitly setting sys.path in the consuming code with an extra directory pointing at these files,
+# it would be possible to generate the files from/to anywhere as modules that did not depend on the dashboard module namespace.
+
+
+# Setting paths for required files and locations
+set(CEPH_DASHBOARD_GATEWAY_PROTO_FILE dashboard/services/proto/gateway.proto)
+set(CEPH_DASHBOARD_PARENT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/../../..")
+set(CEPH_NVME_GATEWAY_PROTO_FILE ${CEPH_DASHBOARD_PARENT_DIR}/../../../src/nvmeof/gateway/control/proto/gateway.proto)
+
+# Current versions of the grpc_tools.protoc module per OS
+set(UBUNTU_22_04_GRPC_TOOLS_VERSION "3.5.1")
+set(CENTOS_9_GRPC_TOOLS_VERSION "3.14.0")
+set(FEDORA_40_GRPC_TOOLS_VERSION "3.19.6")
```

Comment:
Rather than explicitly hard-coding distro-versions (this is going to complicate maintenance, and excludes distros-versions: F41, Ubuntu 24.04, ...). We should only distinguish by breaking changes libproto versions:
- `<3.12`: `optional` not supported (remove `optional` from proto file).
- `>=3.12`: add `--experimental_allow_proto3_optional` (from `3.14` it's not required, but it's still accepted, so let's simplify and add it until some future `3.*` breaks if this is present).

Reply:
They are hardcoded in the variable name only, and you are right it was the bit I least liked about it but it helped me navigate the minefield.
If you follow the logic a bit, below you'll see the F41 and higher are supported but just marked as Unknown.
They can definitely be renamed to match the libproto version.

Regarding the `optional` keyword and `<3.12` ...
There's a reason why I'm doing that, the current version of grpc_tools shipped by Ubuntu22.04 can't handle the current `gateway.proto` as is. 

I'll simplify in line with the suggestions.